Add dependency injection and integration test infrastructure

ajantrania opened the pull request:
This PR introduces dependency injection pattern to enable proper integration testing. Refactors authentication and database connection handling to use injectable providers instead of global state, allowing both production and test implementations.

Key Changes

- Dependency Injection Framework: New AppState with pluggable AuthProvider and ResourcesDbFactory
- Auth Provider Pattern: Extracted authentication logic to RealAuthProvider (production) and FixedAuthProvider (tests)
- Database Factory Pattern: Connection creation now injectable via ResourcesDbFactory trait
- Test Infrastructure: Comprehensive test utilities in tests/common/ with fixtures, helpers, and test router setup
- Integration Tests: Initial health check and report API tests demonstrating the testing approach
- API Changes: router() now async, returns AuthedAccount (account + db connection) from middleware
- Test Support Feature: test-support feature flag exports internal types for testing

TODO:
- [ ] I'm evaluating adding more tests before marking it ready for review.


NOTE: Self-service workflow not tested as I don't have that working yet. Would appreciate close review on that code path in particular.

ajantrania committed:
Add dependency injection and test infrastructure

Introduce DbProvider and AuthProvider traits to enable integration
testing with in-memory databases and injectable auth providers.

- Add DbProvider trait with ProductionDbProvider implementation
- Add AuthProvider trait (RealAuthProvider and FixedAuthProvider)
- Introduce AppState to manage DB and auth context
- Add test-support feature flag for clean API boundaries
- Add integration test infrastructure with test helpers and fixtures

ajantrania committed:
Improve Devcontainer Config

* Fix volume ownership
* Seperate backend cargo registry cache
* Enable port forwarding for local agent<->backend workflow

txase commented on src/db.rs:
```diff
@@ -149,12 +150,26 @@ async fn get_concurrent_db_connection(url: &str) -> anyhow::Result<Surreal<Any>>
+            #[cfg(feature = "rocksdb")]
+            DBConnection::Nonconcurrent(_) => {
+                panic!(
+                    "Cannot clone Nonconcurrent DBConnection (contains MutexGuard). This should never happen in normal operation as Nonconcurrent connections should not be used with Axum Extensions."
```
Since this should never happen, consider using `unreachable!()` instead of `panic!()`? It's a little more idiomatic and informs the code reader what to expect.

txase commented on tests/README.md:
```diff
@@ -0,0 +1,361 @@
+cargo test -p archodex-backend --test '*'
+
+# Run specific test
+cargo test -p archodex-backend test_middleware_loads_account
```
Can we add a launch config task for this for debugging a test, possibly asking the user for input on the test to run and with `-- --nocapture`?

txase commented on tests/report_with_auth_test.rs:
```diff
@@ -0,0 +1,229 @@
+    // Seed test API key in resources database (middleware validates key exists)
+    common::seed_test_api_key(&resources_db, key_id).await;
+
+    // Create FixedAuthProvider that bypasses authentication
```
`s/bypasses authentication/allows api key/`?

txase commented on src/state.rs:
```diff
@@ -0,0 +1,68 @@
+pub trait ResourcesDbFactory {
+    async fn create_accounts_connection(&self) -> anyhow::Result<DBConnection>;
+
+    async fn create_resources_connection(
```
There shouldn't be a `service_url` value when the `archodex-com` feature is disabled. The [previous code](https://github.com/Archodex/archodex-backend/blob/8dc25b8db5ac024da60ec99a64d0432bd98a58b5/src/account.rs#L125-L140) didn't evaluate a `service_url` when the feature is disabled.

I think we should do one of the following:
1. Make the parameter `archodex-com` only:
  ```rust
  async fn create_resources_connection(
      &self,
        account_id: &str,
        #[cfg(feature = "archodex-com")] service_url: Option<&str>,
    ) -> anyhow::Result<DBConnection>;
  ```
2. Add a call to `unreachable!()` if `service_url.is_some()` in the implementation

ajantrania replied:
When I looked into option 1 (which I agree looks cleaner), I ran into a limitation with async-trait and cfg-gated parameters. While [PR #227](https://github.com/dtolnay/async-trait/pull/227) added basic support for cfg on parameters, [Issue #243](https://github.com/dtolnay/async-trait/issues/243) discusses the complications when lifetimes are involved, that is, the macro can't apply cfg to generated lifetime parameters, causing signature mismatches. 

Async-trait has decided not to support this pattern (option 5), recommending cfg-gating within the function body instead. So instead of duplicating the function, I'll opt to using `unreachable!()` to provide a runtime safety check while keeping the signature consistent.

txase reviewed (changes requested):
This looks good to me! Given that we mock out auth and DB connections, I assume you've validated the real version continues to function as expected via some manual testing.

I commented about one thing I saw related to self-hosted, though I don't think it would have broken anything. I expect self-hosted should continue to work just fine with these changes 🤞.

I don't have any any major concerns, just a couple minor things to take a look at.

ajantrania committed:
DX: Improved VS Code test tasks

ajantrania committed:
feat: Add dependency injection for accounts database

Refactor database access to also use DI instead of global state, enabling testing

ajantrania committed:
Fix: Address Env race issue

Address a bug where there was a race condition between the two tests due to the api key being cached. This adds an explicit mechanism to clear the api key cache for tests, and serializes the two problematic tests

ajantrania committed:
WIP: Enable testing for archodex-com mode

Increase Rust stack size to support SurrealDB requirements

ajantrania committed:
WIP: Enable testing for archodex-com mode

Add test to verify serializion of Account struct in both modes

ajantrania committed:
WIP: Enable testing for archodex-com mode

Updating existing integration tests to work or be skipped, as appropraite, based on the archodex-com feature flag

When testing archodex-com, test initialization function will setup required environment variables (ENDPOINT, ACCOUNTS_SUUREALDB_URL)

In archodex-com mode, adding workaround for bypassing call to archodex_com::api_private_key() and instead return a test key to avoid interfacing with KMS

ajantrania committed:
feat: Add unit testing for surrealdb deserializer

ajantrania committed:
WIP: Enable testing for archodex-com mode

Fix test support and test to not break when building for archodex-com

ajantrania commented:
Updates made so far:

* Addressed feedback so far

**Even more Dependency Injection**
* Extended DI pattern for the accounts DB
* DB access now fully injectable via `DbProvider` trait

**Testing enhancements**
* Addressed race condition in tests due to environment variables and cached API private key
* Added SurrealDB deserializer unit tests
* Improved VS Code tasks to provide better DX when running tests locally

**Archodex-com mode testing (WIP)**
🚧 Note: Tests in this mode are not yet fully functional 🚧  
Once this work is complete, I'll squash the WIP commits into a single clean commit.

* Making progress enabling tests to run in archodex-com mode
* Added workaround for KMS during testing by using a pre-defined test key (not in love with this implementation though, but it works for now)
* Tests now adapt based on the mode 
* Increased Rust stack size to accommodate SurrealDB requirements in test scenarios
